### PR TITLE
Update rubocop.yml

### PR DESCRIPTION
**Pull request by damiendillon**
Many of the cops have changed since this was last updated. 

This updates them and also adds new cops aligning it with Devex BE guild agreements.

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -1,1102 +1,394 @@
...
-  Description: >-
-                  Use `tr` instead of `gsub` when you are replacing the same
-                  number of characters. Use `delete` instead of `gsub` when
-                  you are deleting characters.
-  Reference: 'https://github.com/JuanitoFatas/fast-ruby#stringgsub-vs-stringtr-code'
-  Enabled: true
-
-##################### Rails ##################################
+#
+# Rails Cops
```
Comment (dsager) [inline .rubocop.yml]: We should not include any rails cops, as the `gem` does not depend on rails and can be used in non-rails environments.

File: .rubocop.yml [hunk 2]
```diff
@@ -1,1102 +1,394 @@
...
-  Enabled: false
-
+  EnforcedStyle: exploded
+Style/RedundantArrayConstructor:
+  # Excluded so that Grape param declarations, such as:
+  #   type: Array[Integer]/Array[String]/type: Array[Symbol]
+  # do not have to be disabled individually inline, as they would otherwise trigger this cop
+  Exclude:
+    - 'app/controllers/**/*'
+    - 'app/helpers/**/*'
```
Comment (dsager) [inline .rubocop.yml]: As this is a gem, there is no `app` folder

File: .rubocop.yml [hunk 3]
```diff
@@ -1,1102 +1,394 @@
...
-  Enabled: false
-
-Style/GuardClause:
-  Description: 'Check for conditionals that can be replaced with guard clauses'
-  StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals'
-  Enabled: false
-
+  # Built-in global variables are allowed by default.
+  AllowedVariables:
+    - $flipper
```
Comment (dsager) [inline .rubocop.yml]: the gem does not have a dependency on Flipper, so I'd remove this

File: .rubocop.yml [hunk 4]
```diff
@@ -1,1102 +1,394 @@
...
-
-Style/EmptyLiteral:
-  Description: 'Prefer literals to Array.new/Hash.new/String.new.'
-  StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#literal-array-hash'
+  # do not require top level docs for classes/modules
+  Enabled: false
+Style/DocumentationMethod:
+  Exclude:
+    # do not require method documentation in migration files
+    - 'db/migrations/*.rb'
```
Comment (dsager) [inline .rubocop.yml]: the gem doesn't have a `db` folder.

File: .rubocop.yml [hunk 5]
```diff
@@ -1,1102 +1,394 @@
...
+  EnforcedStyle: to_not
+RSpec/SpecFilePathFormat:
+  CustomTransform:
+    RuboCop: rubocop
+    RSpec: rspec
+    OAuthService: oauth_service
+    OAuthController: oauth_controller
+    API: ''
+  Exclude:
+    - 'spec/lib/devex/api/**/*'
```
Comment (dsager) [inline .rubocop.yml]: I think this can be removed

**Comment by damiendillon** [comment]
I can remove everything but I left it all as was aligning it with the idea that we would have a shared rubocop file that we spoke about a long while back.

**Comment by dsager** [comment]
> I can remove everything but I left it all as was aligning it with the idea that we would have a shared rubocop file that we spoke about a long while back.

While this makes sense for our internal applications, I think open source gems are a different use case and the rubocop configuration should be tailored for it.

File: .rubocop.yml (hunk 3 above)
Comment (damiendillon) [reply .rubocop.yml]: Removed.

File: .rubocop.yml (hunk 4 above)
Comment (damiendillon) [reply .rubocop.yml]: Removed.

File: .rubocop.yml (hunk 5 above)
Comment (damiendillon) [reply .rubocop.yml]: Removed.

File: .rubocop.yml (hunk 2 above)
Comment (damiendillon) [reply .rubocop.yml]: Removed.

File: .rubocop.yml (hunk 1 above)
Comment (damiendillon) [reply .rubocop.yml]: Removed.

**Review by ozgun — APPROVED** [review]

**Review by dsager — APPROVED** [review]

**Pull request merged**